[로또] 안돌 미션 제출합니다.

andole98 opened the pull request:
로또 티켓과 로또 생성기를 인터페이스로 뽑아내려다가,
때가 되면 고치는것도 나쁘지 않을 것 같아서 되돌렸습니다.
많은 지적 바랍니다.

andole98 committed:
feat: README.md

로또 게임 규칙 기술.
필요 기능 정리.

andole98 committed:
feat: Lotto

로또 한장을 수행할 Lotto.

andole98 committed:
feat: Lotto Override equals() and hashCode()

로또 객체를 비교한다.

andole98 committed:
feat: UserLottos

유저가 가질 로또.
복수의 로또를 가질 수 있다.

andole98 committed:
feat: LottoGenerator

스태틱 필드로 로또 숫자들을 가지고 있다.
lotto()로 새 로또를 생성후 반환한다.

andole98 committed:
refactor: Lotto.validate()

생성자에 유효성 검사 로직을 부여했다.
validate()는 세 가지 경우를 검사한다.
1. 로또 숫자의 개수가 적절하지 않을 때
2. 로또 숫자가 중복될 때
3. 로또 숫자가 범위를 벗어날 때

andole98 committed:
feat: WinningLotto

match(): 해당 로또와 비교해서 몇개 일치하는지 개수를 반환.

andole98 committed:
feat: Reward

로또 배당을 관리하는 enum.
매칭 개수에 따라 enum을 반환한다.
money()는 enum의 값을 평가한다.

andole98 committed:
refactor: 당첨 확인을 WinnigLotto으로 추출

andole98 committed:
feat: README.md 예외사항 추가

andole98 committed:
feat: OutputView 생성

구입한 로또 출력.
결과 출력.

andole98 committed:
feat: toString 오버라이드

출력을 위한 toString 오버라이드.
리팩토링 시작.

andole98 committed:
refactor: README.md 수정

리팩토링 과제를 구체화.

andole98 committed:
feat: Reward toString 구현

결과 출력을 위해 toString 오버라이드.

andole98 committed:
feat: 보너스 볼 추가

테스트 필요.

andole98 committed:
feat: 보너스 볼 검사

Reward enum에 boolean 추가.

andole98 committed:
feat: LottoResult 생성

로또 결과를 관리할 클래스.
Generator를 붙일 가능성 높음.

andole98 committed:
feat: LottoResult 생성

UserLottos가 결과를 생성.

andole98 committed:
feat: LottoResult.toString() 재작성

OutputView에서 쓰일 포맷

andole98 committed:
refactor: 원시값 상수로 추출

andole98 committed:
feat: Custom Exception 생성

andole98 committed:
issue: 인터페이스

andole98 committed:
Revert "issue: 인터페이스"

This reverts commit 2ee6336b4aeb2ecf468962561c38ccad0045c5ab.

andole98 committed:
feat: ConsoleMessages enum 생성

콘솔에 출력할 메시지들을 enum으로 관리.

andole98 committed:
refactor: 상수값 추출

andole98 committed:
style: Import 정리

andole98 committed:
처음부터 다시

andole98 committed:
feat: UserLottoDto 생성

역할
문자열로 받은 로또 관련 정보를 파싱한다.
UserLottos의 초기화에 필요한 정보를 제공한다.

andole98 committed:
feat: WinninngLottoDto 생성

WinningLotto 생성에 필요한 정보를 파싱하고 전달한다.

andole98 committed:
feat: Interface Ticket 생성

복권 한장에 대한 인터페이스.
Lotto가 구현할 예정.
Lotto는 패키지에 담아두고 Ticket 인터페이스만 나머지에 공개한다.

andole98 committed:
feat: TicketCreator Interface 생성

Lotto는 패키지에 감춰져 있으므로, 로또를 생성해줄 인터페이스 생성.
Lotto에 필요한 유효성 검사 등을 호출한다.
WinningLotto에도 유효성 검사가 필요하므로 제3의 클래스에게 유효성 검사를
의뢰할 가능성이 높다.

andole98 committed:
feat: winningTicket Interface 생성

사용자가 구매한 로또와 비교할 로또.
일치 숫자에 따라 보상이 정해지는 Rank와는 분리하기 위해
일치 숫자와 보너스만을 반환한다.

andole98 committed:
feat: Lotto 생성

default modifier.
세부 구현은 컨트롤러 등에게서 숨긴다.

andole98 committed:
feat: LottoNumbers, LottoNumberManager 생성

Lotto에 사용할 숫자들 관련 헬퍼 클래스.
Ticket 뿐 아니라 WinningTicket에도 사용할 예정.
TicketCreator, Ticket이 주로 호출한다.

andole98 committed:
feat: LottoGenerator 생성

TicketCreator를 구현한 public 클래스.
외부 패키지에서 초기화, 사용될 클래스.

andole98 committed:
feat: AbstractWinningLotto 생성

WinningLotto가 LottoNumberManager의 메서드를 이용해야 한다.
LottoNumberManager의 접근 한정자는 default이므로 패키지 외부에서 접근할
수 없다. WinninTicket을 구현한 Abstract를 패키지 안에 두고
구체 클래스 WinningLotto는 패키지 밖으로 빼서 해결했다.

andole98 committed:
feat: UserLottos 생성

사용자가 구매한 로또를 관리할 클래스.
Ticket컬렉션을 관리한다.
LottoGenerator를 구성한다.
LottoResult를 생성할 책임을 진다.

andole98 committed:
feat: WinningLotto 생성

당첨로또를 관리할 클래스.
유저측 로또와 공통 부분은 TicketModel 패키지 내부 AbstractWinningLotto에
있다.

andole98 committed:
feat: Rank enum 생성

로또 등수에 따른 보상을 관리할 클래스.
보너스유무를 enum 필드로 넣었다.
toString 오버라이드 예정.

andole98 committed:
feat: LottoResult 생성

로또 게임의 결과를 관리할 클래스.
Map을 래핑한다.

andole98 committed:
feat: 테스트 코드 작성

추가 예정... :fire:

andole98 committed:
feat: ConsoleLottoMain 생성

콘솔 진입점 생성.

andole98 committed:
refactor: InputView 역할 변경

Dto를 리턴하지 않고 입력만을 전달한다.
컨트롤러가 Dto 생성후 모델에 전달한다.

andole98 committed:
refactor: README.md 수정

andole98 committed:
refactor: LottoNumberDuplicateException 제거

LottoNumberException으로 통합.

andole98 committed:
fix: Rank.FIRST 오타 수정.

FIST -> FIRST

andole98 committed:
refactor: Lotto class Access Modifier default

andole98 committed:
refactor: AbstractWinningLotto default로 변경

andole98 committed:
test: 테스트코드 추가

UserLotto 테스트 추가 예정.

andole98 committed:
test: UserLottoDto 테스트

콘솔 입력값이 적절히 파싱되는지 확인.

andole98 committed:
feat: Rank toString 오버라이드

Rank를 설명할 수 있게 변경.

andole98 committed:
refactor: Ticket 인터페이스에 toString 추가

andole98 committed:
refactor: LottoNumbers.numbers() 변경

새로운 배열을 리턴하도록 변경.

andole98 committed:
refactor: Lotto.toString() 변경

접두어 제거하고 숫자만 출력하게 변경.

andole98 committed:
refactor: LottoNumberManager.validate 제거

사용하지 않는 메서드 제거.

andole98 committed:
refactor: LottoResult.toString 오버라이드

통계결과도 함께 출력하도록 변경.

andole98 committed:
fix: UserLottos 반복문 내 버그 해결

자동 생성 티켓의 Upbound가 1 많던 것을 수정.
수동 로또가 없을 경우, null값이 들어온다.
iterate 대신 size()대로 반복하도록 바꾸었다.

andole98 committed:
fix: InputView manualLottoNumber 버그 수정

수동로또가 0 일 경우, 수동로또번호를 묻지 않게 변경했다.

andole98 committed:
refactor: Web 컨트롤러 설정 변경

templates 폴더를 루트 디렉토리로 설정.

woowahan-pjs commented on src/main/java/lotto/domain/Exceptions/LottoNumberException.java:
```diff
@@ -0,0 +1,11 @@
+package lotto.domain.Exceptions;
+
+public class LottoNumberException extends LottoException {
```
커스터마이징 한 예외 클래스 👍

woowahan-pjs commented on src/main/java/lotto/domain/Exceptions/LottoNumberException.java:
```diff
@@ -0,0 +1,11 @@
+package lotto.domain.Exceptions;
+
+public class LottoNumberException extends LottoException {
```
`LottoException`을 상속하는 것은 어떤 의미가 있을까요?

andole98 replied:
최상위 컨트롤러가 LottoException을 캐치하게 하려고 했습니다.
예외 발생시 되돌아가는 로직을 굳이 구현하지 않아서 코드에서는 드러나지 않았습니다....

woowahan-pjs replied:
어떤 형태의 코드든 본인이 코드에 대한 믿음을 가지고 일관성 있게 프로그래밍을 한다면 문제없을 거예요. 👍

woowahan-pjs commented on src/main/java/lotto/view/ConsoleMessages.java:
```diff
@@ -0,0 +1,19 @@
+package lotto.view;
+
+public enum ConsoleMessages {
```
View에서 사용하는 메시지를 한곳에 잘 모아놓은 것 같아요. 👍 글로벌 서비스를 위한 다국어 지원, 여러 개의 데이터베이스 연결 등의 요구 사항을 충족하기 위해 종종 사용하는 방법이기도 합니다. Controller와 Model에서 같이 사용하지 않도록 조심하면 되겠어요.

woowahan-pjs commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,55 @@
+    private static final Scanner scanner = new Scanner(System.in);
+
+
+    public static String lottoMoney() {
```
Controller에서 구입 금액을 검증할 수 있도록 int가 아닌 String을 반환하네요! 💯

woowahan-pjs commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,45 @@
+    @Override
+    public String toString() {
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append("당첨통계\n").append("---------\n");
```
Model에 View의 요구 사항을 반영하는 것이 맞을까요? Model과 View의 기능을 분리하면 어떤 장점이 있을까요?

woowahan-pjs commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,45 @@
+            sumOfRank += entry.getKey().money() * entry.getValue();
+            sumOfTickets += entry.getValue();
+        }
+        return sumOfRank / sumOfTickets / LOTTO_MONEY;
```
같은 뜻이지만 다른 개발자도 쉽게 이해할 수 있도록 바꿀 수 있을까요?

```java
return sumOfRank / (sumOfTickets * LOTTO_MONEY);
```

woowahan-pjs commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,45 @@
+            sumOfRank += entry.getKey().money() * entry.getValue();
+            sumOfTickets += entry.getValue();
+        }
+        return sumOfRank / sumOfTickets / LOTTO_MONEY;
```
아래의 글을 읽고 코드에 반영하면 어떨까요?
https://isooo.github.io/java/2019/01/17/BigDecimal.html

andole98 replied:
해당 로직은 int값들을 더하고 double로 나누고 있습니다.
BigDecimal로 처리하려면
BigDecimal 객체 3개가 필요합니다.
일회성 결과를 위해 객체 3개 생성 보다는,
OutputView에서 String.format()을 하는게 더 나아 보입니다.
굳이 필요하다면 OutputView에서, 리턴받은 summury double을 BigDecimal로 변환 후 리턴하면 될 것 같습니다.

woowahan-pjs replied:
현재 요구 사항에서 `sumOfRank`와 `sumOfTickets`, `LOTTO_MONEY`가 정수가 아닌 유리수일 가능성이 없네요. 일반적인 프로그램에서 큰 영향이 없는 미묘한 차이일지라도 결제 또는 정산 프로그램에서는 치명적인 오류를 발생시킬 수 있습니다. `int` 혹은 `long`으로 관리하는 경우, `1_000` 등을 곱해서 계산하기도 합니다.

woowahan-pjs commented on src/main/java/lotto/domain/UserLottos.java:
```diff
@@ -0,0 +1,62 @@
+    @Override
+    public String toString() {
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append(tickets.size()).append("장 구매했습니다.\n");
```
마찬가지로 Model에 View의 요구 사항을 반영하는 것이 맞을까요?

woowahan-pjs commented on src/main/java/lotto/domain/TicketModel/Lotto.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.Objects;
+
+class Lotto implements Ticket {
+    private final List<Integer> numbers;
```
`Number`는 클래스로 포장하여 사용해 볼 수 있겠어요.

andole98 replied:
여기서 `Number`라는 것은 `numbers` 변수가 아니라 `Integer`를 말씀하시는 건가요?

woowahan-pjs replied:
`Integer`를 `LottoNumber`로 잘 포장하셨어요. 👍

woowahan-pjs commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,55 @@
+            iterate = Integer.parseInt(manualCount);
+        } catch (NumberFormatException e) {
+            throw new LottoTicketException();
+        }
```
View에서 꼭 해야 하는 작업일까요?

woowahan-pjs commented on src/main/java/lotto/domain/UserLottos.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    private void generateAuto(int buyMoney) {
+        for (int i = tickets.size(); i < (buyMoney / 1000); i++) {
```
반복 문의 조건식 중 `buyMoney / 1000`에 대한 결과 값을 지역 변수에 담아 사용하는 것과 그렇지 않은 것은 어떤 차이가 있을까요?

woowahan-pjs commented on src/main/java/lotto/domain/Dto/UserLottoDto.java:
```diff
@@ -0,0 +1,51 @@
+package lotto.domain.Dto;
```
패키지 이름은 모두 소문자입니다.

> Package names are all lowercase, with consecutive words simply concatenated together (no underscores). For example, `com.example.deepspace`, not `com.example.deepSpace` or `com.example.deep_space`.

woowahan-pjs replied:
`UserLottoDto`는 `domain` 패키지에 있는 것이 적절할까요? DATA TRANSFER OBJECT는 비즈니스 로직을 가지는 것이 아닌 서로 다른 영역 간의 데이터 이동을 위한 용도로 사용합니다. 외부로부터 데이터의 안전성을 확보할 수도 있습니다. 아래의 글은 가볍게 참고만 해주세요.
https://www.slipp.net/questions/93

woowahan-pjs commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,45 @@
+    FIFTH(3, false, 5000),
+    LOSE(2, false, 0);
+
+
```
공백 라인을 의미 있게 사용하는 것이 좋아 보이며, 문맥을 분리하는 부분에 사용하는 것이 좋습니다. 과도한 공백은 다른 개발자에게 의문을 줄 수 있어요.

woowahan-pjs commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,45 @@
+
+    public static Rank rank(int match, boolean bonus) {
+        return Arrays.stream(values())
+                .filter(rank -> rank.match == match && rank.bonus == bonus)
```
파이프라인을 최대한 나누면 코드의 가독성이 올라갑니다.

woowahan-pjs commented on src/main/java/lotto/domain/TicketModel/AbstractWinningLotto.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.util.List;
+
+public class AbstractWinningLotto implements WinningTicket {
```
`AbstractWinningLotto`라는 클래스명을 보았을 때, 추상 클래스를 의도한 것일까요?

woowahan-pjs commented on src/main/java/lotto/domain/TicketModel/LottoGenerator.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.*;
+
+public class LottoGenerator implements TicketCreator {
```
인터페이스명과 클래스명은 `Creator`와 `Generator` 중 하나를 선택하여 일관성 있게 프로그래밍을 하는 것이 좋아요.

woowahan-pjs commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,11 @@
+import lotto.domain.Dto.WinningLottoDto;
+import lotto.domain.TicketModel.*;
+
+public class WinningLotto extends AbstractWinningLotto {
```
`AbstractWinningLotto`을 상속하는 것은 어떤 의미가 있을까요?

woowahan-pjs commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Map;
+import java.util.TreeMap;
+
+public class LottoResult {
```
`LottoResult`에 대한 테스트 코드가 필요해 보여요.

woowahan-pjs commented on src/main/java/lotto/domain/TicketModel/LottoNumberManager.java:
```diff
@@ -0,0 +1,43 @@
+        for (int i = MIN_NUMBER; i <= MAX_NUMBER; i++) {
+            numbers.add(i);
+        }
+    }
```
인스턴스를 미리 만들어 놓거나 새로 생성한 인스턴스를 캐싱 하여 재활용하는 식으로 불필요한 객체 생성을 피할 수 있다. 👍

woowahan-pjs commented on src/main/java/lotto/domain/TicketModel/LottoNumberManager.java:
```diff
@@ -0,0 +1,43 @@
+    private static final int MIN_NUMBER = 1;
+    private static final int MAX_NUMBER = 45;
+    private static final int NUMBER_COUNT = 6;
+    private static final List<Integer> numbers;
```
마찬가지로 `Number`는 클래스로 포장하여 사용해 볼 수 있겠어요.

andole98 replied:
매니저는 LottoNumbers를 검사하거나 적절한 LottoNumbers를 생성해 주는 역할을 합니다.
그래서 LottoNumbers의 유효성 검사를 LottoNumberManager에게 요청하게 됩니다.
매니저는 숫자들을 최종 관리할 책임이 있으므로, LottoNumbers로 포장된 데이터를 갖고 있으면 안됩니다.
반대로 LottoNumbers가 LottoNumberManager에 의존성을 가집니다. 생성자의 인자로 들어온 값들이 적절한지 매니저에게 물어보죠.
수동과 자동 숫자의 유효성 검증 로직이 분산되는게 좋지 않아서 이렇게 설계했습니다.
어쨋든 제 생각엔 숫자들과 가장 가깝게 있는 객체인 LottoNumberManager가 List<Integer>필드를 갖는게 적절할 것 같습니다.

woowahan-pjs replied:
`LottoNumbers`가 아닌 `LottoNumber`에 대한 코멘트였습니다. "`LottoNumberManager`는 `LottoNumber`를 가진다."라는 설계는 어떤가요?

andole98 replied:
LottoNumber는 Integer 원시값을 말씀하시는 건가요? 잘 이해가 되지 않습니다. ㅠㅠ

woowahan-pjs replied:
`Integer`를 `LottoNumber`로 잘 포장하셨어요. 👍

woowahan-pjs commented on src/main/java/lotto/domain/TicketModel/AbstractWinningLotto.java:
```diff
@@ -0,0 +1,34 @@
+    private final int bonus;
+
+    public AbstractWinningLotto(List<Integer> numbers, int bonus) {
+        TicketCreator generator = new LottoGenerator();
```
`AbstractWinningLotto`가 `LottoGenerator`에 의존하고 있어요. 의존성 주입(dependency injection, DI)이라는 것을 찾아보고 아래의 질문에 답해보면 좋을 것 같아요.

- 의존성 주입(dependency injection, DI)이란 무엇일까요?
- 그것을 하는 이유는 무엇일까요?
- 의존성 주입을 `AbstractWinningLotto`에는 어떻게 적용할 수 있을까요?

andole98 replied:
추상 클래스를 만든 이유는 티켓모델 패키지의 로직을 외부로부터 감추고 싶어서 그랬습니다.
굳이 추상클래스를 확장하게 하기보다 바로 WinningTicket을 구현하는 것으로 바꾸었습니다.

woowahan-pjs reviewed (changes requested):
스스로 고민을 많이 한 흔적이 남아 있는 코드였어요. 👍 
아래의 글이 도움 되면 좋겠어요.

- http://codingnuri.com/seven-virtues-of-good-object
- https://www.slipp.net/questions/253

그 외에도 몇 가지 의견을 남겨놓았으니 충분히 고민하고 도전해보세요!
진행 과정에서 궁금한 점은 댓글이나 Slack을 통해 물어보세요.

andole98 committed:
refactor: dto 패키지 변경

패키지 이름 Dto -> dto
도메인 패키지에서 제외.

andole98 committed:
refactor: LottoResult.toString 변경

LottoResult를 받아 출력하는 OutputView에서 출력물을 조립하게 변경.

andole98 committed:
refactor: LottoResult.summury 변경

public으로 전환했다. OuputView에서 summury값을 받아 가공한다.

andole98 committed:
refactor: WinningLotto 계층구조 변경

AbstractWinningTicket을 삭제.
WinningLotto 가 직접 WinningTicket을 implements 하도록 변경.

andole98 committed:
refactor: Ticket 필드를 LottoNumbers로 추출

List<Integer>를 LottoNumbers로 변경.
필요했던 List.contains()를 LottoNumbers가 구현.

andole98 committed:
refactor: Ticket의 단위 가격 메서드 추가

로또 한장 가격이 여러 도메인에서 필요하다.
Ticket에서 Unit Price를 알려줄 수 있는 메서드를 주고,
Lotto에서 구현하게 하면 public하게 알수 있으면서
Lotto를 가릴 수 있다.

andole98 committed:
refactor: OutputView를 위한 toString 변경

OutputView는 model의 메서드로 값을 get 하도록 바꾸었다.

andole98 committed:
style: 공백 라인 제거

woowahan-pjs commented on src/main/java/lotto/domain/Exceptions/LottoNumberException.java:
```diff
@@ -0,0 +1,11 @@
+package lotto.domain.Exceptions;
```
아래 코멘트에 대해 고민해봐요.
https://github.com/woowacourse/java-lotto/pull/38#discussion_r291855905

andole98 replied:
저번 피드백이었는데.. 죄송ㅠㅠ

woowahan-pjs commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,30 @@
+        for (Map.Entry<Rank, Integer> entry : lottoResult.results().entrySet()) {
+            stringBuilder.append(entry.getKey().toString()).append(" - ").append(entry.getValue()).append(" 개\n");
+        }
+        stringBuilder.append("총 수익률은 ").append(lottoResult.summury()).append("입니다.\n");
```
View는 Model의 주요 기능(`summury()`)을 실행하는 것이 아닌 화면을 보여주는 데에 집중하면 어떨까요?

andole98 committed:
test: LottoResult 테스트

plus(), results() 테스트.

andole98 committed:
feat: 예외 메시지를 관리할 ExceptionMessages

andole98 committed:
refactor: 사용하지 않는 생성자 제거

Map을 넣어주는 생성자 제거. 테스트에 쓰려고 했으나 불필요.

andole98 committed:
feat: LottoNumber 생성

로또 숫자를 담당할 클래스. int 값을 포장한다.
LottoNumbers는 LottoNumber의 컬렉션.
LottoNumberPool은 자동 로또에 사용됨.

woowahan-pjs commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public Map<Rank, Integer> results() {
+        return new TreeMap<>(results);
```
새로운 컬렉션 초기화 👍

woowahan-pjs commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Arrays;
+
+public enum Rank {
+    FIRST(6, false, 2000000000),
```
Java7부터는 `100_000_000`처럼 숫자에 밑줄을 사용하여 가독성을 높일 수 있어요.

woowahan-pjs commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,45 @@
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append(match).append("개 일치 ");
+        if (bonus) {
+            stringBuilder.append("보너스 볼 일치 ").append(money);
```
아래 코멘트에 대해 고민해봐요. 만약 요구 사항에서 보너스 볼 일치를 문자가 아닌 "○" 혹은 로또 공 그림으로 표현하기로 하였다면 어떻게 될까요? 또는 콘솔 화면은 "○", 웹 페이지는 로또 공 그림으로 둘 다 제공하기로 결정되었다면요? Model은 View의 요구 사항에 따라 계속 바뀌는 것이 맞을까요? 쉽지 않겠지만 고민해보고 코드로 표현하면 좋을 것 같아요!
https://github.com/woowacourse/java-lotto/pull/38#discussion_r291853258

woowahan-pjs commented on src/test/java/lotto/domain/TicketModel/LottoCreatorTest.java:
```diff
@@ -0,0 +1,45 @@
+//        LottoCreator generator = new LottoCreator();
+//        LottoNumbers numbers = new LottoNumbers(Arrays.asList(1, 2, 3, 4, 5, 6));
+//        assertEquals(new Lotto(numbers), generator.create(Arrays.asList(1, 2, 3, 4, 5, 6)));
+//    }
```
프로그래밍을 하다 보면 다음과 같은 느낌이 들어 코드를 삭제하지 않고 주석 처리하는 경우가 있는데요.

- 이 코드가 지금은 사용하지 않지만 언젠가 사용할 수 있을 것 같다.
- 이 코드를 내가 정말 힘들게 구현해서 삭제하기 아깝다.
- 분명 사용하지 않는 코드로 파악되지만 내가 구현한 코드가 아니라 함부로 삭제하기 힘들다.

과거에는 Subversion과 Git 같은 형상관리 툴을 사용하지 않았고, 코드를 참고할만한 곳이 많지 않았기 때문에 주석을 통해 유지하는 것이 의미가 있었을 거예요. 하지만 최근 개발 환경은 형상관리 툴을 사용하는 것이 일반적이 되어 이전에 구현한 코드를 손쉽게 원복 할 수 있게 되었어요. 오히려 지금은 코드를 얼마나 깔끔하고 가독성이 좋은 코드로 유지하는지가 더 중요하지 않을까요? 현시점에 굳이 필요 없다고 생각하는 코드가 있다면 아까운 마음을 버리고 가차 없이 삭제해보세요.

woowahan-pjs commented on src/test/java/lotto/domain/LottoResultTest.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @Test
+    void 외부에서_맵_변경() {
```
좋은 학습 테스트네요. 👍 학습 테스트란 기능 구현을 위한 테스트라기 보다 새로운 API, 라이브러리, 프레임워크가 어떻게 동작하는지를 검증하기 위한 테스트이며, 아래와 같은 장점이 있습니다.

- 다양한 조건에 따른 기능을 손쉽게 확인해 볼 수 있다.
- 학습 테스트 코드를 개발 중에 참고할 수 있다.
- 프레임워크나 제품을 업그레이드할 때 호환성 검증을 보여준다.
- 테스트 작성에 대한 좋은 훈련이 된다.
- 새로운 기술을 공부하는 과정이 즐거워진다.

woowahan-pjs commented on src/test/java/lotto/domain/TicketModel/LottoCreatorTest.java:
```diff
@@ -0,0 +1,45 @@
+//    }
+
+    @Test
+    void 매칭확인() {
```
`LottoCreator`에 대한 테스트가 맞을까요?

woowahan-pjs commented on src/test/java/lotto/domain/TicketModel/LottoNumberPoolTest.java:
```diff
@@ -0,0 +1,12 @@
+class LottoNumberPoolTest {
+    @Test
+    void 자동생성() {
+        assertNotEquals(new LottoNumbers(), new LottoNumbers());
```
`LottoNumberPool`을 테스트하기 위해서 `LottoNumbers`에 대한 테스트가 꼭 통과해야 할까요?

woowahan-pjs commented on src/test/java/lotto/domain/TicketModel/LottoNumbersTest.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Test
+    void 자동생성() {
+        assertNotEquals(new LottoNumbers(), new LottoNumbers());
```
"매번 새로운 인스턴스가 만들어진다."에 대한 테스트일까요?

woowahan-pjs commented on src/main/java/lotto/domain/UserLottos.java:
```diff
@@ -0,0 +1,58 @@
+    private final TicketCreator ticketCreator;
+    private List<Ticket> tickets;
+
+    public UserLottos(UserLottoDto dto) {
```
`UserLottos`를 생성하기 위해서는 `UserLottoDto`가 필요하네요. View의 요구 사항(콘솔 화면과 웹 페이지에서 받고자 하는 값의 형태 등)이 바뀐다면 Model도 그것에 영향을 받지 않을까요? Model은 외부의 영향을 받지 않도록 만드는 것이 좋습니다.

woowahan-pjs commented on src/main/java/lotto/dto/UserLottoDto.java:
```diff
@@ -0,0 +1,52 @@
+        }
+    }
+
+    private List<List<Integer>> parseNumber(List<String> manualNumbers) {
```
`UserLottoDto`는 서로 다른 영역 간의 데이터 이동을 위한 용도 이상으로 사용되고 있네요.

woowahan-pjs commented on src/main/java/lotto/domain/TicketModel/LottoNumberPool.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.Collections;
+import java.util.List;
+
+public class LottoNumberPool {
```
수동 생성인 경우에도 `LottoNumberPool`를 사용할 수 있도록 해보세요.

woowahan-pjs commented on src/main/java/lotto/domain/TicketModel/LottoNumberPool.java:
```diff
@@ -0,0 +1,29 @@
+        for (int i = LottoNumber.MIN_NUMBER; i <= LottoNumber.MAX_NUMBER; i++) {
+            ticketNumbers.add(new LottoNumber(i));
+        }
+    }
```
인스턴스를 미리 만들어 놓거나 새로 생성한 인스턴스를 캐싱 하여 재활용하는 식으로 불필요한 객체 생성을 피할 수 있다. 👍

woowahan-pjs commented on src/main/java/lotto/ConsoleLottoMain.java:
```diff
@@ -0,0 +1,33 @@
+            manualCount = Integer.parseInt(InputView.manualLottoCount());
+        } catch (NumberFormatException e) {
+            throw new LottoTicketException(ExceptionMessages.TICKET.message());
+        }
```
아래의 코드는 의미를 설명할 수 있는 메서드로 한 번 더 분리하는 것이 좋아 보여요.

```java
try {
    manualCount = Integer.parseInt(InputView.manualLottoCount());
} catch (NumberFormatException e) {
    throw new LottoTicketException(ExceptionMessages.TICKET.message());
}
```

woowahan-pjs reviewed (changes requested):
원시 값 포장과 테스트 코드 추가 등 피드백을 잘 반영해주셨습니다. 👍 
설계한 것을 소리 내어 읽어보고 말이 되는지 확인해보는 것이 많은 도움이 된답니다.
고민해보면 좋을 것 같은 의견을 남겨놓았으니 충분히 고민하고 도전해보세요!
다음 리뷰 요청 때는 Merge 하도록 하겠습니다.

andole98 committed:
style: 패키지 이름 변경

소문자로 변경.

andole98 committed:
style: 금액 표시법 변경

천단위 구분.

andole98 committed:
style: 필요 없는 주석 제거

andole98 committed:
test: LottoNumberPool 테스트 추가

andole98 committed:
feat: ConsoleLottoMain.toManualLottoCount() 메서드 생성

View로부터 받은 입력을 숫자로 변환.

andole98 committed:
feat: LottoResultDto 생성

View에 전달하기 위한 Dto.

andole98 committed:
refactor: LottoNumberPool, LottoCreator  역할 변경.

Lotto를 이루는 LottoNumbers와 같은 객체는
final 필드를 초기화 할 수 있는 생성자 하나만 가진다.
Lotto는 LottoNumberPool과 관계하지 않고 Creator가 관계한다.

andole98 committed:
refactor: 티켓 비교 로직 변경

Ticket -> TicketNumbers -> TicketNumber까지 깊이 들어가지 않고
Ticket에서 처리할 수 있게 바꿈.

andole98 committed:
refactor: LottoResult.result() 반환 형식 변경

필드를 그대로 리턴하지 않고 Dto를 리턴한다.

andole98 committed:
feat: LottoResultDto 생성

OutputView에서 사용할 Dto.

andole98 committed:
feat: UserLottoSeed 생성

UserLottos 생성자 로직 변경.
UserLottoSeed가 생성에 필요한 정보를 넘겨준다.

andole98 committed:
feat: WinningLottoSeed 생성

WinningLotto 생성자 로직 변경.
WinningLottoSeed가 생성자 인자를 만들어준다.

andole98 committed:
feat: Ticket toString() 생성

Lotto, LottoNumbers, LottoNumber toString() 생성

andole98 committed:
refactor: OutputView.printResult() 변경

LottoResultDto를 사용하도록 변경.

andole98 committed:
test: 테스트 케이스 수정

andole98 committed:
refactor: README.md 수정

andole98 committed:
style: LottoResultDto 상수 추출

소수점 반올림 자릿수 3으로 추출.

andole98 committed:
refactor: UserLottos 필드 제거

LottoResult 지역 변수로 변경.

andole98 committed:
refactor: Rank.toString() 변경

OuputView에서 Rank 값에 따라 메시지를 조합하도록 변경.
Rank의 콘솔 메시지 제공 책임은 OutputView로 이전.

andole98 commented:
오프라인 피드백이 정말 큰 도움이 되었습니다. 감사합니다.

woowahan-pjs reviewed (changes requested):
오프라인 코드 리뷰로 대신하였습니다. 👍 
아래의 글이 도움 되면 좋겠어요.

- https://blog.naver.com/muchine98/220304821784

andole98 committed:
refactor: UserLottos.result() 변경

LottoResultDto가 UserLottos를 참조하도록 변경.

andole98 committed:
refactor: core 패키지 변경

외부 컨트롤러에 직접 드러나지 않는 패키지로 기준 변경.

andole98 committed:
feat: ResultSummary 인터페이스 생성

LottoResult의 통계 처리를 담당할 타입.
AverageSummary는 평균 수익률로 통계를 낸다.

andole98 committed:
feat: UserLottosCreator 생성

UserTicket를 생성한다.

andole98 committed:
feat: WinningLottoCreator 생성

WinningTicket을 생성한다.

andole98 committed:
style: import 패키지명 변경

andole98 committed:
refactor: ConsoleLottoMain 변경

타입으로 객체를 관리.

andole98 committed:
test: 테스트코드 변경

andole98 committed:
refactor: 패키지 정리

andole98 commented:
이번 오프라인 피드백 역시 큰 도움이 되었습니다. 감사합니다.
인터페이스가 과하게 사용된 느낌이 큽니다. 세부 구현체는 하나인 경우가 많은데, 어떻게 생각하시나요?

woowahan-pjs reviewed (changes requested):
오프라인 코드 리뷰로 대신하였습니다. 👍 
아래의 글이 도움 되면 좋겠어요.

- http://woowabros.github.io/study/2018/03/05/sdp-sap.html

andole98 committed:
refactor: 패키지 구조 변경

service 패키지 추가.

andole98 committed:
refactor: 인터페이스 제거, 패키지 정리

service 패키지 생성.

woowahan-pjs reviewed (approved):
축하합니다! 101개의 Commit이 작성되었어요. 🥇 
다음 단계를 진행하려면 woowacouse에 Merge 된 코드에서 다시 시작해야 충돌이 발생하지 않을 거예요.
아래의 과정을 진행한 후, 다음 단계 진행할 것을 추천해요.

```
1. git remote add upstream
2. git fetch upstream 자신의 github_id
3. git reset --hard upstream/자신의 github_id
```

다음 단계에서 뵙겠습니다. 이만 Merge 하겠습니다.
